tools/generators/files_and_groups/src/Generator/Elements/Elements+File.swift

```diff
@@ -0,0 +1,117 @@
+import PBXProj
+
+extension Generator.Elements {
+    private static let folderTypeFileExtensions: Set<String?> = [
+       "bundle",
+       "docc",
+       "framework",
+       "scnassets",
+       "xcassets",
+       "xcdatamodel",
+    ]
+
+    /// Creates a `PBXFileReference` element.
+    static func file(
+        node: PathTreeNode,
+        parentBazelPath: BazelPath,
+        specialRootGroupType: SpecialRootGroupType?,
+        collectBazelPaths: Generator.Environment.CollectBazelPathsFunction,
+        elementAttributes: Generator.Environment.Elements.AttributesFunction,
+        elementIdentifier: Generator.Environment.Elements.IdentifierFunction,
+        elementHashCache: inout Set<String>
+    ) -> (
+        element: Element,
+        bazelPaths: [BazelPath],
+        resolvedRepository: ResolvedRepository?
+    ) {
+        let bazelPath = parentBazelPath + node
+
+        var contentComponents = [
+            "{isa = PBXFileReference;",
+        ]
+
+        let ext = node.extension()
+
+        let lastKnownFileType: String?
+        let sortOrder: Element.SortOrder
+        if node.isFolder && !folderTypeFileExtensions.contains(ext) {
+            lastKnownFileType = "folder"
+            sortOrder = .groupLike
+        } else {
+            lastKnownFileType = ext.flatMap { Xcode.filetype(extension: $0) }
+            sortOrder = .fileLike
+        }
+        if let lastKnownFileType {
+            contentComponents.append(
+                "lastKnownFileType = \(lastKnownFileType.pbxProjEscaped);"
+            )
+        }
+
+        let explicitFileType: String?
+        if node.name == "BUILD" {
+            explicitFileType = Xcode.filetype(extension: "bazel")
+        } else if node.name == "Podfile" {
```

Comment:
Do we need/want to also handle the lockfile here?

Reply:
I don't think people will add the lockfile to `xcodeproj.extra_files`. This was added as part of #2040.

Reply:
Cool that makes sense